module/applications/components/damage-application.mjs

```diff
@@ -0,0 +1,334 @@
+          <div class="target-source-control">
+            <button type="button" class="unbutton" data-mode="targeted" aria-pressed="true">
+              <i class="fa-solid fa-bullseye" inert></i> ${game.i18n.localize("DND5E.Tokens.Targeted")}
+            </button>
+            <button type="button" class="unbutton" data-mode="selected" aria-pressed="false">
+              <i class="fa-solid fa-expand inert"></i> ${game.i18n.localize("DND5E.Tokens.Selected")}
+            </button>
+          </div>
+          <ul class="targets unlist"></ul>
+          <button class="apply-damage" type="button" data-action="applyDamage">
+            <i class="fa-solid fa-reply-all fa-flip-horizontal inert"></i>
+            ${game.i18n.localize("DND5E.Apply")}
+          </button>
+        </div>
+      `;
+      this.replaceChildren(div);
+      this.applyButton = div.querySelector(".apply-damage");
+      this.applyButton.addEventListener("click", this._onApplyDamage.bind(this));
+      this.targetList = div.querySelector(".targets");
+      this.targetSourceControl = this.querySelector(".target-source-control");
+      this.targetSourceControl.querySelectorAll("button").forEach(b =>
+        b.addEventListener("click", this._onChangeTargetMode.bind(this))
+      );
+      if ( !game.user.isGM || !this.chatMessage.getFlag("dnd5e", "targets")?.length ) {
+        this.targetSourceControl.hidden = true;
+      }
+      div.querySelector(".collapsible-content").addEventListener("click", event => {
+        event.stopImmediatePropagation();
+      });
+    }
+
+    this.buildTargetsList();
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Build a list of targeted tokens based on current mode & replace any existing targets.
+   */
+  async buildTargetsList() {
+    // Ensure collapsible section automatically resizes when content changes
+    const collapsible = this.querySelector(".collapsible-content");
+    if ( collapsible.style.height !== "0px" ) collapsible.style.height = "auto";
+
+    let targetedTokens;
+    switch ( this.targetingMode ) {
+      case "targeted":
+        targetedTokens = (this.chatMessage.getFlag("dnd5e", "targets") ?? []).map(t => t.uuid);
+        if ( this.selectedTokensHook ) {
+          Hooks.off("controlToken", this.selectedTokensHook);
+          this.selectedTokensHook = null;
+        }
+        break;
+      case "selected":
+        targetedTokens = canvas.tokens.controlled.map(t => t.document.actor.uuid);
+        this.selectedTokensHook ??= Hooks.on("controlToken", foundry.utils.debounce(() => this.buildTargetsList(), 50));
+        break;
+    }
+    const targets = await Promise.all(targetedTokens.map(t => this.buildTargetListEntry(t)));
+    if ( targets.length ) this.targetList.replaceChildren(...targets.filter(e => e));
+    else {
+      const li = document.createElement("li");
+      li.classList.add("none");
+      li.innerText = game.i18n.localize("DND5E.Tokens.NoneTargeted");
```

Comment:
I guess for polish this should probably say 'no tokens targeted' or 'no tokens selected', depending on the current target mode.